File: 백준/Gold/11054. 가장 긴 바이토닉 부분 수열/가장 긴 바이토닉 부분 수열.py
n = int(input())
d = list(map(int, input().split()))
dp_l = [1]*n
dp_r = [1]*n
for i in range(n):
    for j in range(i):
        if d[j] < d[i] :
            dp_l[i] = max(dp_l[j]+1, dp_l[i])
for i in reversed(range(n)):
    for j in range(n-1, i-1, -1):
        if d[j] < d[i] :
            dp_r[i] = max(dp_r[j]+1, dp_r[i])
dp = [dp_l[i]+dp_r[i] for i in range(n)]
print(max(dp)-1)

File: 백준/Gold/11054. 가장 긴 바이토닉 부분 수열/가장 긴 바이토닉 부분 수열.java
import java.io.IOException;
import java.util.Arrays;

public class Main {
    static int[] data, dp_l, dp_r, dp;
    static int n;

    public static void main(String[] args) throws IOException {
        n = readInt();
        data = new int[n];
        for (int i = 0; i < n; i++) {
            data[i] = readInt();
        }

        System.out.println(solution());
    }

    private static int solution() {
        bi_l();
        bi_r();
        dp = new int[n];
        for (int i = 0; i < n; i++) {
            dp[i] = dp_l[i] + dp_r[i];
        }

        return Arrays.stream(dp).max().getAsInt() - 1;

    }

    private static void bi_l() {
        dp_l = new int[n];
        for (int i = 0; i < n; i++) {
            dp_l[i] = 1;
            for (int j = 0; j < i; j++) {
                if (data[j] < data[i]) {
                    dp_l[i] = Math.max(dp_l[i], dp_l[j] + 1);
                }
            }
        }
    }

    private static void bi_r() {
        dp_r = new int[n];
        for (int i = n - 1; i > -1; i--) {
            dp_r[i] = 1;
            for (int j = n - 1; j > i; j--) {
                if (data[j] < data[i]) {
                    dp_r[i] = Math.max(dp_r[i], dp_r[j] + 1);
                }
            }
        }
    }

    static int readInt() throws IOException {
        int v, t = 0;
        while ((v = System.in.read()) > 47) {
            t = t * 10 + (v - 48);
        }
        if (v == 13) {
            System.in.read();
        }
        return t;
    }
}